[OpenACC] Reduction 'init' lowering for all-ones/least/largest

erichkeane opened the pull request:
As a follow on to the last patches of this form, this patch does the init section for all of the reduction operators that weren't previously covered, which is '&' as all-ones, 'max' as 'least', and 'min' as 'largest'.

erichkeane committed:
[OpenACC] Reduction 'init' lowering for all-ones/least/largest

As a follow on to the last patches of this form, this patch does the
init section for all of the reduction operators that weren't previously
covered, which is '&' as all-ones, 'max' as 'least', and 'min' as
'largest'.

andykaylor commented on clang/lib/Sema/SemaOpenACC.cpp:
```diff
@@ -2589,26 +2589,63 @@ SemaOpenACC::ActOnOpenACCAsteriskSizeExpr(SourceLocation AsteriskLoc) {
+  case InitKind::AllOnes:
+    return llvm::APFloat::getAllOnesValue(Context.getFloatTypeSemantics(Ty));
+  case InitKind::Least:
+    return llvm::APFloat::getSmallestNormalized(
```
Are you sure you don't want `llvm::APFloat::getLargest(Context.getFloatTypeSemantics(Ty), /*Negative=*/true);` here? I think `getSmallestNormalized` is going to return a value near zero.

erichkeane replied:
Ah!  Good to know, thank you!

andykaylor commented on clang/test/CIR/CodeGenOpenACC/combined-reduction-clause-float.cpp:
```diff
@@ -131,9 +150,28 @@
+// CHECK-NEXT: cir.store{{.*}} %[[LARGEST]], %[[DECAY]] : !cir.float, !cir.ptr<!cir.float>
+// CHECK-NEXT: %[[ONE_IDX:.*]] = cir.const #cir.int<1> : !s64i
+// CHECK-NEXT: %[[NEXT_ELT:.*]] = cir.ptr_stride(%[[DECAY]] : !cir.ptr<!cir.float>, %[[ONE_IDX]] : !s64i), !cir.ptr<!cir.float>
+// CHECK-NEXT: %[[LARGEST:.*]] = cir.const #cir.fp<3.4{{.*}}> : !cir.float
```
Maybe also check for the `E+38` here?

erichkeane replied:
Thats a good idea, I'm going to try to be more specific on all the FP checks other than 0, 1, all-ones.

andykaylor commented on clang/test/CIR/CodeGenOpenACC/combined-reduction-clause-float.cpp:
```diff
@@ -142,9 +180,28 @@
-// TODO OpenACC: Expecting an initialization to... SOME value here.
+// CHECK-NEXT: %[[ALLOCA:.*]] = cir.alloca !cir.array<!cir.float x 5>, !cir.ptr<!cir.array<!cir.float x 5>>, ["openacc.reduction.init", init]
+// CHECK-NEXT: %[[DECAY:.*]] = cir.cast(array_to_ptrdecay, %[[ALLOCA]] : !cir.ptr<!cir.array<!cir.float x 5>>), !cir.ptr<!cir.float>
+// CHECK-NEXT: %[[LEAST:.*]] = cir.const #cir.fp<-{{.*}}> : !cir.float
```
```suggestion
// CHECK-NEXT: %[[LEAST:.*]] = cir.const #cir.fp<-3.4{{.*}}E+38> : !cir.float
```

andykaylor commented on clang/lib/Sema/SemaOpenACC.cpp:
```diff
@@ -2627,16 +2664,40 @@ static Expr *GenerateReductionInitRecipeExpr(ASTContext &Context,
+        Exprs.push_back(FloatingLiteral::Create(
+            Context, getInitFloatValue(Context, IK, EltTy),
+            /*isExact=*/true, EltTy, ExprRange.getBegin()));
+        Exprs.push_back(FloatingLiteral::Create(
```
This probably isn't what you want. For example, for `InitKind::One` this will generate `1+1i` but the identity value for complex multiplication is `1 + 0i`. Maybe best to leave it as an error-producing case.

erichkeane replied:
Could you guide me as to what the right answer is for them?  I could definitely split these up into separate functions, but I'd love to know what the '0', '1', All-Ones, Least and Largest should be here?

I think I'd one day soon want to enable complex here.

erichkeane replied:
Ah, hmm.. ok then, I think leaving complex as ill-formed is the best idea. 

I DID just realize my test before was wrong, and I HAVEN'T been disallowing complex, so I'm going to go through and do that as a RAC patch.

andykaylor replied:
Oops. Magnitude is an absolute value, so I guess you'd need {0, 0} for the LEAST case. Anyway, it's probably just a bad idea to allow it at all.

erichkeane replied:
I could probably make complex legal for some operators (+/*?), but it doesn't seem worth the effort until someone asks for it.  In the meantime, I've made sure it is ill-formed always, here: 7a0dfb1fec20bcb8565b39507e282b12e436ba1f

erichkeane committed:
set floats to all something reasonable/nonsense

erichkeane committed:
Do a bunch of review stuff Andy suggested:

-Fix 'largest negative' for floats.
-Make complex just emit 'zero', we are going to make it ill-formed, but
it should produce a valid initializer at least
-update all tests to properly check exponent for floats too.

andykaylor reviewed (approved):
lgtm
